block/manager.go

```diff
@@ -101,9 +101,27 @@ type Manager struct {
 
 // getInitialState tries to load lastState from Store, and if it's not available it reads GenesisDoc.
 func getInitialState(store store.Store, genesis *cmtypes.GenesisDoc) (types.State, error) {
+	// Load the state from store.
 	s, err := store.GetState()
 	if err != nil {
+		// If the user is starting a fresh chain (or hard-forking), we assume the stored state is empty.
 		s, err = types.NewFromGenesisDoc(genesis)
+	} else {
+		// Perform a sanity-check to stop the user from
+		// using a higher genesis than the last stored state.
+		// if they meant to hard-fork, they should have cleared the stored State
+		if uint64(genesis.InitialHeight) > s.LastBlockHeight {
```

Comment:
i don't think it is correct.  the supplied `genesis.InitialHeight` is expected to be `s.LastBlockHeight` + 1, because `InitialHeight` refers to the height at which you want to start creating blocks. e.g., for empty chain, `InitialHeight` is 1, not 0.

Reply:
i see now. thanks for the clarification.